aries_vcx/src/protocols/did_exchange/service/responder/response_sent/mod.rs

```diff
@@ -0,0 +1,89 @@
+use messages::msg_fields::protocols::did_exchange::{complete::Complete, response::Response};
+
+use crate::{
+    errors::error::{AriesVcxError, AriesVcxErrorKind},
+    protocols::did_exchange::{
+        service::helpers::create_our_did_document,
+        states::{completed::Completed, responder::response_sent::ResponseSent},
+        transition::{transition_error::TransitionError, transition_result::TransitionResult},
+    },
+};
+
+use self::{
+    config::ReceiveRequestConfig,
+    helpers::{construct_response, resolve_their_ddo},
+};
+
+use super::DidExchangeServiceResponder;
+
+pub mod config;
+mod helpers;
+
+impl DidExchangeServiceResponder<ResponseSent> {
+    pub async fn receive_request(
+        ReceiveRequestConfig {
+            wallet,
+            resolver_registry,
+            request,
+            service_endpoint,
+            routing_keys,
+            invitation_id,
+        }: ReceiveRequestConfig,
+    ) -> Result<TransitionResult<DidExchangeServiceResponder<ResponseSent>, Response>, AriesVcxError> {
+        let their_ddo = resolve_their_ddo(&resolver_registry, &request).await?;
+        let (our_ddo, enc_key) = create_our_did_document(&wallet, service_endpoint, routing_keys).await?;
+
+        if request.decorators.thread.and_then(|t| t.pthid) != Some(invitation_id.clone()) {
+            return Err(AriesVcxError::from_msg(
+                AriesVcxErrorKind::InvalidState,
+                "Parent thread ID of the request does not match the id of the invite",
+            ));
+        }
+
+        let response = construct_response(our_ddo, invitation_id.clone(), request.id.clone())?;
+
+        Ok(TransitionResult {
+            state: DidExchangeServiceResponder::from_parts(
+                ResponseSent {
+                    request_id: request.id,
+                    invitation_id,
+                },
+                their_ddo,
+                enc_key,
+            ),
+            output: response,
+        })
+    }
+
+    pub fn receive_complete(
+        self,
+        complete: Complete,
+    ) -> Result<DidExchangeServiceResponder<Completed>, TransitionError<Self>> {
+        if complete.decorators.thread.thid != self.state.request_id {
```

Comment:
I know this mirror mostly current approach and we've been tracking ids/threads in state machines, it's been questioned whether this should really be happening, and whether the state machines shall do a bit less - only care for the protocol part, but less so for decorators.
I am not sure if that's always the case, and whether it's possible to assume a state machine would *never* need access to some decorator. 

I do tend to lean toward the idea that we should deal with thids, ids, pthids on protcol level. Either way, to look up and attempt to update a state machine, upper layers of code will definitely have to do it the same thing - inspect thid/pthid/id of a arriving message and use that information to lookup the proper state machine in storage in order to perform update. So checking thid, pthids etc. in transitions ultimately leads to duplication which already must been performed by consumer of this code.
I guess this could make again make SMs further bit more lightweight in terms of logic and content of the states.

(And I wonder if doing so could by a chance lead to further to rethinking any other high level design decisions in terms of protocols/state machines)

Reply:
Fair enough, there's not reason to make contract, on the other hand, from all protocols we deal with today, at their core, each of them can work without using decorators whatso-ever. So asking for full message, including decorators, when you actually only need the "protocol piece" to make things work (eg. doing something what'd be considered driving state/protocol forwrad/generate reply). 

But that links to the other point - currently we indeed access decorators. So the second point I argue for, is that we probably should remove `thid / pthid / id` verifications on state machine level. What we are actually doing is runtime testing of the code consuming this. 
While it's nice to have and - "why not to make assert if you can" - it doesn't quite feel inlined with single resposibility principle. 

Btw I know some 2 years ago I claimed the opposite - that state machine should have awareness of its "identity", by knowing thread it belongs to - as it's essential piece of information. Back than we had philosophy of keeping serialization model backwards compatible, and the idea was that you will ALWAYS end up wanting linking a FSM with thread id, and since we provide stable serialization format, we can do that for the user. 
Obviously we change our thoughts on that, and serialized state machines are "black boxes" now - so users of aries-vcx are definitely (and we know from our experience) tracking threads ids on their own. Hence that explains why I changed position on this. I now think state machines shouldn't know/care "who they are", and should be burden of the upper layer to assure passing in correct messages.

Reply:
One thing I have missed is consideration that without dealing with `thread_id`, `id` etc. we can't build complete "ready-to-send" response msgs.
But I can see that - along with other decorators like `~timing`, localization etc. - being often part of an upper layer, some message processor in the future. Otherwise, if keep inflating state machine / transition layer with additional logic (set up timing decorator, set up locatization decorator and what not, ... ) either end up being very opinionated, or we further expand function argument with config which will direct the logic how the decorators shall be set up (and the function arguments are already quite massive, albeit wrapped into a struct)

But it can be gradual process. We can start building full blown messages now, then pull some logic (attaching decorators) on an upper "framework" layer later. 
All that being said, I still think we should probably start in a point where state machine doesn't remember it's `thid`.

Reply:
As mentioned in private, I have many reservations to what you said there (but perhaps am aligned with the way you are going). I don't think imposing assumptions left and right is not a good approach (even when it might lead to a good solution). I would much rather think in terms of components, responsibilities and data, but for that, we need to first use common language, and you are using terms for which we might have different definitions in confusing way (using term "state machine" while commenting on `service` module, etc.).

Ultimately, I think that GitHub comments are not the best medium for this conversation.